### PR TITLE
Support custom Vault secret prefix

### PR DESCRIPTION
**Pull request by sam-heilbron**
# Description

Allow injecting a custom `pathPrefix` for Vault secrets, historically this was hard-coded to "secret".

# Context

## Why is this configuration valuable?

It allows for users to use custom secrets engines and for multitenancy-ness for organizations with different/separated secrets paths.

## How did this work before?

This value was previously hardcoded in solo-kit https://github.com/solo-io/solo-kit/blob/1d799ae290c2f516f01fc4ad20272d7d2d5db1e7/pkg/api/v1/clients/vault/resource_client.go#L309

## How did we verify this?

* [Created a test](https://github.com/solo-io/gloo/pull/6283/files#diff-0b902b4e3694a4bc0293894f70516d1f3ed3ee5710f44ae08adeabfdc729feccR70) to verify that our vault client can interact with the custom `--vault-path-prefix` fed in by user. The test can create a secret using the custom path prefix, and can also read from Vault.
* We [updated our test setup](https://github.com/solo-io/gloo/pull/6283) to allow for the creation of a custom `-path` for the Vault test server to test the pathPrefix. This matches how Vault customers enable new secrets engines with custom path.

# Checklist:

- [X] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [X] If I updated APIs (our protos) or helm values, I ran `make -B install-go-tools generated-code` to ensure there will be no code diff
- [X] I followed guidelines laid out in the Gloo Edge [contribution guide](https://docs.solo.io/gloo-edge/latest/contributing/)
- [X] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/6184

### CODE REVIEW

File: projects/gloo/pkg/bootstrap/vault.go [hunk 1]
```diff
@@ -4,8 +4,19 @@ import (
 	"github.com/hashicorp/vault/api"
 	errors "github.com/rotisserie/eris"
 	v1 "github.com/solo-io/gloo/projects/gloo/pkg/api/v1"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients/factory"
 )
 
+const DefaultPathPrefix = "secret"
+
+func NewVaultSecretClientFactory(client *api.Client, pathPrefix, rootKey string) factory.ResourceClientFactory {
```
Comment (sam-heilbron) [inline projects/gloo/pkg/bootstrap/vault.go:12]: We instantiate a VaultSecretClientFactory both in our CLI (to configure secrets) and our Setup (to consume secrets). To ensure that those two are always configured the same way, I opted to introduce this constructor.

File: projects/gloo/pkg/bootstrap/vault.go [hunk 2]
```diff
@@ -4,8 +4,19 @@ import (
 	"github.com/hashicorp/vault/api"
 	errors "github.com/rotisserie/eris"
 	v1 "github.com/solo-io/gloo/projects/gloo/pkg/api/v1"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients/factory"
 )
 
+const DefaultPathPrefix = "secret"
+
```
Comment (nfuden) [inline projects/gloo/pkg/bootstrap/vault.go:11]: nit: golangci-lint makes me suggest the following
```suggestion
// NewVaultSecretClientFactory consumes a vault client along with a set of basic configurations for retrieving info with the client
```

File: projects/gloo/pkg/bootstrap/vault.go [hunk 3]
```diff
@@ -4,8 +4,19 @@ import (
 	"github.com/hashicorp/vault/api"
 	errors "github.com/rotisserie/eris"
 	v1 "github.com/solo-io/gloo/projects/gloo/pkg/api/v1"
+	"github.com/solo-io/solo-kit/pkg/api/v1/clients/factory"
 )
```
Comment (nfuden) [inline projects/gloo/pkg/bootstrap/vault.go:9]:
```suggestion
// The DefaultPathPrefix may be overridden to allow for non-standard vault mount paths 
```

**Comment by inFocus7** [comment]
re-running build, flake #2838

**Comment by sam-heilbron** [comment]
The most recent build-bot failure: `[Fail] Happy path Rest Eds Enabled: (V3) in memory ssl sni [It] should work with ssl` is one we have not seen before. My instinct is to say this is due to test pollution. 

In the logs I noticed a message that seems to occur whenever we see this type of new flake: 
```
Step #10 - "test": WARN	gloo.setup.gloosnapshot.event_loop.envoyTranslatorSyncer	Proxy had invalid config after xds sanitization	{"proxy": "name:\"gateway-proxy\" namespace:\"gloo-system\"", "error": "2 errors occurred:\n\t* invalid resource gloo-system.gateway-proxy\n\t* WARN: \n  [Route Warning: InvalidDestinationWarning. Reason: *v1.Upstream { default.local-test-upstream-70-ssl } not found]\n\n"}
Step #10 - "test": INFO	gloo	respond open watch 1[] with new version "14695981039346656037"
```

I believe that https://github.com/solo-io/gloo/pull/6301 will resolve this issue. I vote that for now we kick this flake and track it. And if I can confirm that this linked PR will resolve the issue then I can close any new flake issues

**Comment by inFocus7** [comment]
had flake #6253

**Comment by inFocus7** [comment]
/kick no fail reason shown, CI build failed during "docker-push-extended": Pushed

**Comment by inFocus7** [comment]
/kick 
```
  Expected
      <core.Status_State>: 1
  to equal
      <core.Status_State>: 2
``` 

kicking due to possible flake in test

**Comment by nfuden** [comment]
/kick Locally the kv test that just failed doesnt. So its a flake... though we should consider how we can fix it if this turns out to be truly a flake

**Comment by inFocus7** [comment]
test flake in **CI / k8s regression tests (glooctl, false) (pull_request)**. Istio/istioctl failed to install, created new issue for it #6316

File: projects/gloo/api/v1/settings.proto [hunk 4]
```diff
@@ -113,6 +113,12 @@ message Settings {
         // this can be used to run multiple instances of Gloo against the same Consul cluster
         // defaults to `gloo`
         string root_key = 9;
+
+        // pathPrefix, if set, represents the name of a Vault Secrets Engine
+        // https://learn.hashicorp.com/tutorials/vault/getting-started-secrets-engines
+        // The path prefix tells Vault which Secrets Engine to which it should route traffic
+        // defaults to 'secret'
```
Comment (Rachael-Graham) [inline projects/gloo/api/v1/settings.proto:120]:
```suggestion
        // Optional: The name of a Vault Secrets Engine to which Vault should route traffic.
        // For more info see https://learn.hashicorp.com/tutorials/vault/getting-started-secrets-engines.
        // Defaults to 'secret'
```

File: projects/gloo/cli/pkg/flagutils/common.go [hunk 5]
```diff
@@ -70,7 +70,8 @@ func AddVaultSecretFlags(set *pflag.FlagSet, vault *options.Vault) {
 
 	set.BoolVar(&vault.UseVault, "use-vault", false, "use Vault Key-Value storage as the "+
 		"backend for reading and writing secrets")
-	set.StringVar(&vault.RootKey, "vault-root-key", bootstrap.DefaultRootKey, "key prefix for for Vault key-value storage.")
+	set.StringVar(&vault.PathPrefix, "vault-path-prefix", bootstrap.DefaultPathPrefix, "The path prefix tells Vault which Secrets Engine to which it should route traffic.")
```
Comment (Rachael-Graham) [inline projects/gloo/cli/pkg/flagutils/common.go:73]:
```suggestion
	set.StringVar(&vault.PathPrefix, "vault-path-prefix", bootstrap.DefaultPathPrefix, "The Secrets Engine to which Vault should route traffic.")
```

**Review by Rachael-Graham — COMMENTED** [review]
doc

**Review by nfuden — APPROVED** [review]
Looks good. Most of the decisions were in the solokit PR.

**Pull request merged**